### ISSUE TITLE
Questions about training.

### ISSUE THREAD

**Issue by FathUMinUllah3797** [IDLabMedia/comprint#3]
# Thanks for publishing a good work.

I have two questions, kindly answer me.

1. After running the training codes, I could not able to see where I can other metrics such as IoU, Accuracy, or TPV/NPV etc. Could you be pleased ?
2. If we want to train the network with other datasets such as CASIA, COLUMBIA, or other image manipulation datasets, should we consider the pristine images only? 
3. Next, after running the training code, the training log shows the loss and val-loss only, how to see other metrics?

**Comment by hmareen** [comment]
Hi Fat U Min Ullah,

Concerning the **metrics** (in your first and third point): during training, we have not used any other metrics than the binary cross entropy loss. Note that the network output is not a forgery classification (real/forged) that is easy to measure using metrics such as the accuracy. Instead, the network outputs a fingerprint. It is only _after_ analyzing the predicted fingerprint, that we can classify pixels as real or forged (using SPAM filters and the EM algorithm). So, we only calculated such measures after training, using image manipulation datasets.

If, however, you wish to define your own metrics and measure these during training, you can add them in the metrics-parameter of the [Model.compile()](https://www.tensorflow.org/api_docs/python/tf/keras/Model#compile) method. In the siamese training code, that call can be found [here](https://github.com/IDLabMedia/comprint/blob/main/code/train_network_siamese.py#L99).

Concerning other **datasets** (in your second point): the strength of Comprint is that is trained on pristine images only. As such, the network cannot overfit on specific manipulation types of the datasets, and should generalize well to a variety of manipulation types. The dataset that we trained on consists of uncompressed images that were compressed in a variety of ways, and we teach the network that fingerprints of images that underwent the same compression should be similar (and those that underwent different compression should be dissimilar).

When using image manipulation datasets that contain both pristine and manipulated images, the network may overfit to these types of manipulations. Using the pristine images only could work if you also compress them in a variety of ways, such that you can teach the network that fingerprints of images that underwent the same compression should be similar. But I am not sure what would be the benefit of this?

The [tampCOCO dataset of CATNet](https://github.com/mjkwon2021/cat-net#1-downloading-tampcoco--compraise) may be a good alternative dataset to train on, as they spliced images of different compressions. However, in that case, we cannot do Siamese training with pairs of images. Instead, we should train the network on a single image and force the same/different fingerprint according to the corresponding mask of that image. That would require writing a new training method, though.

I hope my answers are clear. If not, feel free to ask some more questions.
